### PR TITLE
[e2e] Move migration tests to their own package

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
`migration_test.go` file has grown a lot over time.
This has compromised the readability of the file itself, as well as its maintainability.
In addition, containing several overlapping `Contexts`, several flakes have recently been discovered due to the test setup as much as the code base.
This is the first in a series of PRs whose purpose is to refactor and reorganize migration tests file.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This PR is an attempt to split the commits of https://github.com/kubevirt/kubevirt/pull/9545 into separate PRs, following this path:
- Create `migration` package and other necessary changes.
- Create several tests files based on the context and the functionality.

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: The [Boy Scout Rule](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html) was considered and applied if needed
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to the [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```
/cc @xpivarc @jean-edouard

### CODE REVIEW

**Comment by fossedihelm** [comment]
/hold
Some missing  stuffs

**Comment by fossedihelm** [comment]
> Thank you for the effort, great to see this happening!
> 
> One thing I'd really want to see is the split of `tests/migration/migration.go` into several files (similar to #9902). Are you planning to do this in another PR?

Thanks @dhiller! Yes, I preferred to not include the commits you are referring to in this initial PR to keep the size smaller and limit the contexts.
So yes, it is planned to do the split in a followup PR(s) :slightly_smiling_face:

File: tests/migration/migration_utils.go
Comment (dhiller) [inline tests/migration/migration_utils.go:37]:
```suggestion
func ExpectMigrationToSucceedWithOffset(offset int, virtClient kubecli.KubevirtClient, migration *v1.VirtualMachineInstanceMigration, timeout int) *v1.VirtualMachineInstanceMigration {
```

File: tests/migration/migration_utils.go
Comment (dhiller) [inline tests/migration/migration_utils.go:111]: nit: do not export this one, replace calls of this func to the Set/Clear funcs.

File: tests/swap_test.go
Comment (dhiller) [inline tests/swap_test.go:137]: nit: what the func does is not clear to me by reading it
```suggestion
		nodeAffinityRuleForVmiToFill, err := migration_utils.CreateNodeAffinityRuleToMigrateFromSourceToTargetAndBack(targetNode, targetNode)
```

File: tests/sysprep_test.go
Comment (dhiller) [inline tests/sysprep_test.go:222]: nit: not sure here, maybe `libwindows.Firmware`? But also in the next iteration, maybe.

File: tests/utils.go
Comment (dhiller) [inline tests/utils.go:1147]: The following funcs look like duplicates of [the libinfra funcs](https://github.com/kubevirt/kubevirt/blob/2e6f4848622acce160adb2f8753444a84baccb54/tests/libinfra/metrics.go#L39)

File: tests/vmi_lifecycle_test.go
Comment (dhiller) [inline tests/vmi_lifecycle_test.go:73]: Maybe move to `libvmi` and rename to `AddNodeAffinity`

**Review by dhiller — CHANGES_REQUESTED** [review]
Thank you @fossedihelm for this initiative. Mostly nits here, some change requests inline.

File: tests/utils.go
Comment (fossedihelm) [reply tests/utils.go:1147]: They are leftover and not used functions. Thank you, will drop them.

File: tests/migration/migration_utils.go
Comment (fossedihelm) [reply tests/migration/migration_utils.go:1]: I will create both `libmigration` and `migration` packages. The first one will be for the migration utils, the latter for the tests. WDYT? Thank you

File: tests/migration/migration.go
Comment (fossedihelm) [reply tests/migration/migration.go:321]: Right! Thanks!

File: tests/vmi_lifecycle_test.go
Comment (fossedihelm) [reply tests/vmi_lifecycle_test.go:73]: I was going to do it in a later PR, but I'll add a commit at this point. I am going to drop the entire function, switch using `WithNodeAffinity` Option in `libvmi`

File: tests/sysprep_test.go
Comment (fossedihelm) [reply tests/sysprep_test.go:222]: Not sure about the need of a `libwindows` package. Will leave as is for the moment, and see in the future. Thank you

**Comment by fossedihelm** [comment]
@dhiller Thanks for the detailed review! I addressed the comments! PTAL Thanks

**Review by jean-edouard — COMMENTED** [review]
Looks great overall!
2 comments:
- Let's avoid making `tests/utils.go` longer, see in-line comment
- I guess it comes down to personal preference, but the renames in last commit don't make much sense to me. In fact, I feel like `ExpectMigrationSuccess` reads better than `ExpectMigrationToSucceed`, same for the other ones. The fact that Ginkgo has the `.To(Succeed())` syntax doesn't mean we should use these exact words in our function names... Anyway, just a nit, feel free to ignore!

File: tests/utils.go
Comment (fossedihelm) [reply tests/utils.go]: Thanks! Instead of `NewRandowWindowsVMI` under `tests/utils` I opted for a `NewWindowsVMI` under `libvmi/factory`.
Regarding the move of most/all the utility functions from `utils` to `libvmi` I have some ideas, but I think this is out of the scope of this PR :) Thanks

**Comment by fossedihelm** [comment]
> * I guess it comes down to personal preference, but the renames in last commit don't make much sense to me. In fact, I feel like `ExpectMigrationSuccess` reads better than `ExpectMigrationToSucceed`, same for the other ones. The fact that Ginkgo has the `.To(Succeed())` syntax doesn't mean we should use these exact words in our function names... Anyway, just a nit, feel free to ignore!

It comes from a previously review from Daniel. I will leave as is for the moment if you don't mind. :) Thanks
PTAL

File: tests/hotplug/cpu.go
Comment (dhiller) [inline tests/hotplug/cpu.go:153]: nit: maybe add another method inside `libmigration` that avoids repeating the `libmigration.MigrationWaitTime` in this file three times?

```suggestion
			libmigration.ExpectMigrationToSucceedWithDefaultTimeout(virtClient, migration)
```

File: tests/monitoring/monitoring.go
Comment (dhiller) [inline tests/monitoring/monitoring.go:99]: nit: avoid repeating default migration timeout here and elsewhere

```suggestion
				migration = libmigration.RunMigrationAndExpectToCompleteWithDefaultTimeout(virtClient, migration)
```

File: tests/network/vmi_istio.go
Comment (dhiller) [inline tests/network/vmi_istio.go:211]: nit: sounds like it has similarities in `libmigration`? Maybe extract a func and move it or see whether we can use something from over there?

File: tests/utils.go
Comment (dhiller) [inline tests/utils.go:114]: This looks like a duplicate from what is inside `libmigration`?

File: tests/utils.go
Comment (dhiller) [inline tests/utils.go:771]: nit: in the long term we should remove all the [man-in-the-middle funcs](https://www.refactoring.com/catalog/removeMiddleMan.html) here entirely.

**Review by dhiller — CHANGES_REQUESTED** [review]
Looks great already!

Sorry that I overlooked two things I think valuable, both are around repeated use of `libmigration.MigrationWaitTime`.

File: tests/utils.go
Comment (fossedihelm) [reply tests/utils.go:114]: Rebase leftover! Thank you!

File: tests/network/vmi_istio.go
Comment (fossedihelm) [reply tests/network/vmi_istio.go:211]: I don't have to extract anything, since the function is already present: `RunMigrationAndExpectToComplete`.
Thank you

File: tests/utils.go
Comment (fossedihelm) [reply tests/utils.go:771]: Agree, but IMHO it will be a little out of the scope of this PR. I think it will be great to do it in a followup PR. WDYT? Thanks

**Comment by fossedihelm** [comment]
@dhiller Comments addressed, PTAL Thanks

**Review by jean-edouard — COMMENTED** [review]
/approve

**Comment by brianmcarey** [comment]
@fossedihelm it looks like this will require a rebase after the latest merge.

**Comment by fossedihelm** [comment]
@dhiller rebase only! Can you lgtm again? thanks

**Review by dhiller — APPROVED** [review]

**Pull request merged**
